docs/IE_PLUGIN_DG/NewTransformation.md

```diff
@@ -1,15 +1,390 @@
 
-// ! [new_transformation:part1]
+To decide where to store your transformation code please follow these rules:
+1. If it's plugin specific transformation and can't be reused by other plugins keep source code inside plugin.
+2. If this transformation relates to OpSetXToOpSetY conversion or it's common optimization then keep sources inside transformation library.
 
-And this file must refer to that code via
+After you decided where to store your transformation code you can start develop your own nGraph transformation.
 
-@snippet src/template_transformation.cpp new_transformation:part1
+## Table of Contents:
+
+1. [`ngraph::Function` and graph representation](#ngraph_function) 
+2. [Transformations types](#transformations_types)
+3. [Pattern matching](#pattern_matching)
+4. [Working with ngraph::Function](#working_with_ngraph_function)
+5. [Transformation writing essentials](#transformation_writing_essentials)
+6. [Common mistakes in transformations](#common_mistakes)
+7. [Using pass manager](#using_pass_manager)
+8. [How to debug transformations](#how_to_debug_transformations)
+9. [Disabling/Enabling specific transformations for plugin X](#disabling_transformation)
+10. [Transformations testing](#transformations_testing)
+
+## ngraph::Function and graph representation <a name="ngraph_function"></a>
+
+nGraph function is a very simple thing: it stores shared pointers to `ngraph::op::Result` and `ngraph::op::Parameter` operations that are inputs and outputs of the graph. 
+All other operations hold each other via shared pointers: child operation holds its parent (hard link). If operation has no consumers and it's not Result operation
+(shared pointer counter is zero) then it will be destructed and won't be accessible anymore. Each operation in `ngraph::Function` is a shared_ptr and has `ngraph::Node` as a base class.
+
+Below you can find examples how `ngraph::Function` can be created:
+
+@snippet example_ngraph_utils.cpp ngraph_utils:simple_function
+
+@snippet example_ngraph_utils.cpp ngraph_utils:advanced_function
+
+## Transformations types <a name="transformations_types"></a>
+
+There are two main transformation types:
+
+###1. ngraph::pass::FunctionalPass
+
+ngraph::pass::FunctionalPass is used for transformations that take entire `ngraph::Function` as input and process it.
+
+Template for FunctionPass transformation class
+
+@snippet src/template_function_transformation.hpp function_pass:template_transformation_hpp
+
+@snippet src/template_function_transformation.cpp function_pass:template_transformation_cpp
+
+Using `ngraph::FunctionPass` you need to override `run_on_function` method where you will write transformation code. Return value must be `true` if original function has changed during transformation (new operation were added or operations replacement was made or node attributes were changed) otherwise it must be `false`.For transformation API please follow [working with ngraph::Function](#working_with_ngraph_function) section.
```

Comment:
missing space before `For transformation`

Reply:
done